File: cmo/cmo_bridge.py
"""cmo_bridge — pont fichier Python(hote Linux+3090) <-> CMO(VM Windows), via dossier PARTAGE.
Meme principe que le pont Arma : CMO ecrit state.json (etat des unites), Python ecrit cmd_<N>.lua
(ordres en Lua, numerotes). Ecritures ATOMIQUES (tmp+rename) -> a travers le partage VM, CMO ne lit
jamais un fichier a moitie ecrit (lecon payee sur Arma)."""
import os, json, time, glob, re


class CMOBridge:
    def __init__(self, bridge_dir):
        self.dir = bridge_dir
        os.makedirs(self.dir, exist_ok=True)
        self.n_sent = self._max_cmd_on_disk()

    def _state_path(self):
        return os.path.join(self.dir, "state.json")

    def _max_cmd_on_disk(self):
        m = 0
        for f in glob.glob(os.path.join(self.dir, "cmd_*.lua")):
            mm = re.search(r"cmd_(\d+)\.lua$", f)
            if mm:
                m = max(m, int(mm.group(1)))
        return m

    def read_state(self, timeout=0.0):
        """Renvoie le dernier etat ecrit par CMO (dict) ou None. Tolere une lecture pendant l ecriture."""
        p = self._state_path()
        t0 = time.time()
        while True:
            try:
                with open(p, "r") as f:
                    return json.load(f)
            except (FileNotFoundError, ValueError):
                if time.time() - t0 >= timeout:
                    return None
                time.sleep(0.05)

    def last_executed(self):
        """Numero du dernier ordre que CMO a execute (champ n de state.json)."""
        st = self.read_state()
        return st.get("n", 0) if st else 0

    def send(self, lua_code, wait=True, timeout=10.0):
        """Ecrit un ordre Lua (numerote) que l actuateur CMO executera. Ecriture atomique."""
        self.n_sent += 1
        n = self.n_sent
        tmp = os.path.join(self.dir, ".cmd_%d.tmp" % n)
        dst = os.path.join(self.dir, "cmd_%d.lua" % n)
        with open(tmp, "w") as f:
            f.write(lua_code)
        os.replace(tmp, dst)                       # atomique -> CMO ne voit que le fichier complet
        if wait:
            t0 = time.time()
            while time.time() - t0 < timeout:
                if self.last_executed() >= n:
                    return True
                time.sleep(0.05)
            return False
        return True

    def units(self, side=None):
        """Raccourci : liste des unites du dernier etat, filtrable par camp."""
        st = self.read_state()
        if not st:
            return []
        us = st.get("units", [])
        return [u for u in us if side is None or u.get("side") == side]

File: cmo/cmo_bridge.lua
-- cmo_bridge.lua — PONT CMO <-> Python (jumeau du pont fichier Arma). A poser comme action d'un
-- EVENT RECURRENT (trigger "Regular Time", ~1 s de jeu) dans un scenario. Mapping Proton Z:\ = /.
--   OUT : dump l'etat (unites des sides Blue/Red : lat/lon/cap/vitesse/alt/vivant) -> state.txt
--   IN  : lit cmd.lua (genere par Python), l'execute via load(), puis l'archive (n croissant)
local DIR = "Z:\\tmp\\cmo_bridge\\"

local function dump_state()
  local f = io.open(DIR .. "state.txt", "w")
  if not f then return end
  f:write("T " .. tostring(ScenEdit_CurrentTime()) .. "\n")
  for _, sidename in ipairs({"Blue", "Red"}) do
    local ok, side = pcall(VP_GetSide, {name = sidename})
    if ok and side and side.units then
      for _, ud in ipairs(side.units) do
        local ok2, u = pcall(ScenEdit_GetUnit, {guid = ud.guid})
        if ok2 and u then
          f:write(string.format("U %s|%s|%s|%.5f|%.5f|%.1f|%.1f|%.1f\n",
            sidename, tostring(u.name), tostring(u.guid),
            u.latitude or 0, u.longitude or 0, u.heading or 0, u.speed or 0, u.altitude or 0))
        end
      end
    end
  end
  f:close()
end

local function exec_cmd()
  local cf = io.open(DIR .. "cmd.lua", "r")
  if not cf then return end
  local code = cf:read("*a"); cf:close()
  os.remove(DIR .. "cmd.lua")
  if code and #code > 2 then
    local chunk, e = load(code)
    if chunk then pcall(chunk) end
    local af = io.open(DIR .. "ack.txt", "w")
    if af then af:write("ACK " .. tostring(ScenEdit_CurrentTime()) .. "\n"); af:close() end
  end
end

exec_cmd()
dump_state()
